### PR TITLE
Add widget::EventFilter matching

### PR DESCRIPTION
**Pull request by jplatte**
… and some other small improvements.

### CODE REVIEW

**Review by toger5 — APPROVED** [review]
Looks good.
To get it right, this includes the actual filter code/filter logic with the FFI but intentionally excludes the parsing?

Other than that it looks very much like the version with the parsing which looked over earlier today.

**Comment by jplatte** [comment]
Yes, this intentionally skips parsing because that's specific to the permissions use of event filters, they're not really supposed to be parsed from a string on their own.

File: bindings/matrix-sdk-ffi/src/widget.rs [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+use matrix_sdk::{
+    async_trait,
+    widget::{MessageLikeEventFilter, StateEventFilter},
+};
 
 use crate::room::Room;
 
 #[derive(uniffi::Record)]
 pub struct Widget {
     /// Information about the widget.
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/src/widget.rs:12]: Should we update the comment of the `settings` field?

File: crates/matrix-sdk/src/widget/filter.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
+#![allow(dead_code)] // temporary
```
Comment (Hywan) [inline crates/matrix-sdk/src/widget/filter.rs:1]: Still necessary?

File: crates/matrix-sdk/src/widget/mod.rs [hunk 3]
```diff
@@ -4,22 +4,26 @@ use async_channel::{Receiver, Sender};
...
+    filter::{EventFilter, MessageLikeEventFilter, StateEventFilter},
+    permissions::{Permissions, PermissionsProvider},
+};
 
 /// Describes a widget.
 #[derive(Debug)]
 pub struct Widget {
     /// Information about the widget.
-    pub info: Info,
+    pub settings: WidgetSettings,
```
Comment (Hywan) [inline crates/matrix-sdk/src/widget/mod.rs:19]: Should we update the doc of `settings`?

**Review by Hywan — APPROVED** [review]

File: crates/matrix-sdk/src/widget/filter.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk/src/widget/filter.rs:1]: Yes, until the `matches` function actually starts getting used outside of tests.

**Pull request merged**